### PR TITLE
Branchless ortho-basis construction algorithm

### PR DESCRIPTION
**Pull request by aytekaman**
Reference: http://jcgt.org/published/0006/01/01/

### CODE REVIEW

File: src/appleseed/foundation/math/basis.h [hunk 1]
```diff
@@ -178,57 +178,23 @@ inline void Basis3<T>::build(const VectorType& normal)
...
-            m_u[2] =  T(0.0);
-        }
-    }
-
-    // u is orthogonal to n, but not unit-length. Normalize it.
-    m_u = normalize(m_u);
-
-    // Compute v.
-    m_v = cross(m_u, m_n);
+    T sign = copysignf(1.0f, m_n[2]);
```
Comment (dictoon) [inline src/appleseed/foundation/math/basis.h]: Should be `T(1.0)`.

**Comment by dictoon** [comment]
Could you add unit benchmarks for this, in single and double precision, comparing with the old method (which you can reproduce in the unit benchmark file)?

File: src/appleseed/foundation/math/basis.h (hunk 1 above)
Comment (est77) [inline src/appleseed/foundation/math/basis.h]: and maybe copysign?

**Comment by dictoon** [comment]
It looks like some unit tests fail because some of the assertions checking the orthonormal basis fail. It would be interesting to know how precise or imprecise the new method is.

File: src/appleseed/foundation/math/basis.h (hunk 1 above)
Comment (dictoon) [inline src/appleseed/foundation/math/basis.h]: Right!

Comment (aytekaman) [inline src/appleseed/foundation/math/basis.h]: Thanks! will correct them.

Comment (est77) [inline src/appleseed/foundation/math/basis.h]: I guess we'd can wrap copysign and copysignf into a overloaded function and then the compiler will do the right thing.

Comment (dictoon) [inline src/appleseed/foundation/math/basis.h]: We don't need to, `std::copysign()` is already a bunch of overloads.

The main problem is that it's C++11 only.

Comment (dictoon) [inline src/appleseed/foundation/math/basis.h]: Let's try without copysign for now?

**Comment by aytekaman** [comment]
I just saw the unit test results. I will investigate this further.

**Comment by aytekaman** [comment]
İ just swapped n_u, n_v. Now unit tests don't fail. Maybe their coordinate system assumptions are different?

**Comment by dictoon** [comment]
Interesting. Our coordinate system is right-handed, with U pointing right, N pointing up and V pointing away from the screen, toward the viewer.

File: src/appleseed/foundation/math/basis.h (hunk 1 above)
Comment (aytekaman) [inline src/appleseed/foundation/math/basis.h]: you mean i should use `copysignf` or the branchless version in the paper which doesn't have `copysign` at all?

Comment (dictoon) [inline src/appleseed/foundation/math/basis.h]: I would suggest doing without `copysign` at all for now.

Comment (dictoon) [inline src/appleseed/foundation/math/basis.h]: So,

    const T sign = m_n[2] < 0.0f ? T(-1.0) : T(1.0);

Also, could you please add const qualifiers?

**Comment by aytekaman** [comment]
Now I will add unit tests.

**Comment by dictoon** [comment]
Awesome!

**Comment by aytekaman** [comment]
Here are the results for convenience:
-------------------------------------------------------------------------------------------------
 BranchlessBuild_SinglePrecision:   2,148,606 clock ticks (1.070K calls/s at 2299.415 MHz)
 OriginalBuild_SinglePrecision:       6,410,783 clock ticks (377.4 calls/s at 2419.280 MHz)
-------------------------------------------------------------------------------------------------
 BranchlessBuild_DoublePrecision: 3,963,067 clock ticks (602.1 calls/s at 2386.097 MHz)
 OriginalBuild_DoublePrecision:     6,577,000 clock ticks (380.7 calls/s at 2503.997 MHz)
-------------------------------------------------------------------------------------------------

**Comment by aytekaman** [comment]
Here are the results from the _release_ build for convenience:

```
BranchlessBuild SinglePrecision: 1,942,049 clock ticks (1.194K calls/s at 2318.736 MHz)
OriginalBuild   SinglePrecision: 6,370,353 clock ticks (386.4 calls/s at 2461.768 MHz)

BranchlessBuild DoublePrecision: 4,019,368 clock ticks (581.1 calls/s at 2335.743 MHz)
OriginalBuild   DoublePrecision: 6,548,406 clock ticks (378.6 calls/s at 2479.023 MHz)
```

File: src/appleseed/foundation/meta/benchmarks/benchmark_basis.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2010-2013 Francois Beaune, Jupiter Jazz Limited
```
Comment (dictoon) [inline src/appleseed/foundation/meta/benchmarks/benchmark_basis.cpp]: If this is a new file (it appears so), you should replace these two lines by a single one with your name:

    // Copyright (c) 2017 Aytek Aman, The appleseedhq Organization

The rules are explained here: https://github.com/appleseedhq/appleseed/wiki/Copyright-Notices

File: src/appleseed/foundation/meta/benchmarks/benchmark_basis.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        Vector<T, 3>  m_normal[N];
+        Basis3<T>     m_result[N];
+
+        BuildFixture()
+        {
+            MersenneTwister rng;
+
+            for (size_t i = 0; i < N; ++i)
+            {
+                m_normal[i] = rand_vector1<Vector<T, 3>>(rng);
```
Comment (dictoon) [inline src/appleseed/foundation/meta/benchmarks/benchmark_basis.cpp]: `rand_vector1<Vector<T, 3>>` is only allowed in C++11 and later. In C++03 and earlier, you need to write

    rand_vector1<Vector<T, 3> >

File: src/appleseed/foundation/meta/benchmarks/benchmark_basis.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                m_normal[i] = normalize(m_normal[i]);
+            }
+        }
+
+        Basis3<T> build_Original(const Vector<T, 3> normal) const
```
Comment (dictoon) [inline src/appleseed/foundation/meta/benchmarks/benchmark_basis.cpp]: `build_original` (method names are lower case).

Comment (dictoon) [inline src/appleseed/foundation/meta/benchmarks/benchmark_basis.cpp]: Can this method be made static instead of merely const?

**Comment by dictoon** [comment]
Very nice work, great attention to details, thanks!

Did the performance change after your last commit ("Make comparison more fair")?

**Comment by dictoon** [comment]
(Make sure to show the outdated comments.)

**Comment by aytekaman** [comment]
I forgot to mention that I had updated the results with the newest commit. Before that, branchless code was faster (ex: Single precision branchless version was around 2,2 M clock ticks)

Maybe we can keep the old method in the Basis implementation to make the comparison more reliable. What do you think?

**Comment by dictoon** [comment]
> Maybe we can keep the old method in the Basis implementation to make the comparison more reliable. What do you think?

What you can do is try move it back there in your local repo, and check whether it makes a noticeable performance difference. I suspect it won't, in which case it's fine to keep the old method in the unit benchmark file.

**Pull request merged**